### PR TITLE
feat: add bird export parser and reader

### PR DESCRIPTION
**Pull request by sakateka**
#### Pull Request Description:

This pull request adds significant functionality and improvements to the Yanet2 control plane, focusing primarily on the `route` module and its integration with the Bird routing daemon. Below are the key highlights of the changes:

 -   **Support for Bird Export Protocol:**
        Introduced a new Bird Export Protocol reader (`bird/export.go`) that allows Yanet2 to process dynamic routing updates from Bird. This integration enhances the platform's capability to maintain accurate and up-to-date routing tables.</br>
        Implemented configuration (`bird/cfg.go`) and parsing logic (`bird/parser.go` and `bird/update.go`) to handle Bird's routing data feed, ensuring robust and correct interpretation.

 -   **Improved Configuration Management:**
        Added default configuration functions in various components for easier initialization (`cfg.go`), ensuring consistent setup and deployment of the platform across environments.

 -   **Optimized Data Structures and Handling:**
        Improved the internal handling of routing information with the introduction of new data structures allowing efficient storage and retrieval of routing data.</br>
        Enhanced the `MapTrie` implementation to better manage route updates and deletions, including a new method `UpdateOrDelete`.

 -   **Neighbor and Hardware Route Management:**
        Refined neighbor entry management by introducing a HardwareRoute structure, which encapsulates source and destination MAC addresses in a **hashable** format for more reliable route tracking.

 -   **Testing Enhancements:**
        Comprehensive tests have been added for new functionalities such as Bird protocol parsing and update mechanics. This ensures robustness against edge cases and correct functionality under various scenarios (`bird/cfg_test.go`, `bird/update_test.go`).

 -   **Concurrency and Performance:**
        Revised concurrency patterns to manage route updates and neighbor information more effectively, improving performance in the routing service operations.

 -   **Overall Robustness and Extensibility:**
        These updates position the Yanet2 platform to better handle large-scale and dynamic network environments, leveraging Bird for real-time route management and propagation.

This pull request marks a substantial upgrade to the routing capabilities of the Yanet2, making it more efficient and flexible in responding to dynamic network changes.

### CODE REVIEW

File: controlplane/modules/route/cfg.go [hunk 1]
```diff
@@ -8,7 +13,19 @@ type Config struct {
...
 	// MemoryRequirements is the amount of memory that is required for a single
 	// transaction.
-	MemoryRequirements uint   `yaml:"memory_requirements"`
-	Endpoint           string `yaml:"endpoint"`
-	GatewayEndpoint    string `yaml:"gateway_endpoint"`
+	MemoryRequirements datasize.ByteSize `yaml:"memory_requirements"`
+	Endpoint           string            `yaml:"endpoint"`
+	GatewayEndpoint    string            `yaml:"gateway_endpoint"`
+	// BirdExport configures the reader for the Bird Export Protocol feed.
+	BirdExport *bird.Config `yaml:"bird_export"`
```
Comment (3Hren) [inline controlplane/modules/route/cfg.go:20]: This is optional field, right? What happens if we omit `bird_export` in the config?

File: controlplane/modules/route/internal/discovery/bird/cfg.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		return err
+	}
+	if m.Enable && len(m.Sockets) == 0 {
+		return fmt.Errorf("bird export is enabled but no sockets are provided")
+	}
+	return nil
+}
+
+type Socket struct {
+	Name string `yaml:"name"`
```
Comment (3Hren) [inline controlplane/modules/route/internal/discovery/bird/cfg.go:23]: Please describe what do you expect here and why.

File: controlplane/modules/route/internal/discovery/bird/export.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		sockets: sockets,
+		cfg:     cfg,
+		updater: ribUpdater,
+		log:     log,
+	}
+}
+
+func (m *Export) Run(ctx context.Context) error {
+	if !m.cfg.Enable {
+		m.log.Info("Bird export reader is disabled")
```
Comment (3Hren) [inline controlplane/modules/route/internal/discovery/bird/export.go:53]: Here and everywhere below: please start a logging message with lowercase letter.

File: controlplane/modules/route/internal/discovery/bird/export.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		return nil
+	}
+
+	updates := make(chan *rib.Route, 100)
```
Comment (3Hren) [inline controlplane/modules/route/internal/discovery/bird/export.go:57]: Is `100` is a magic constant or it's some kind proved reasonable value?

File: controlplane/modules/route/internal/discovery/bird/export.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+				zap.String("name", socket.name),
+				zap.String("path", socket.path))
+
+			c, err := net.Dial("unix", socket.path)
+			if err != nil {
+				return fmt.Errorf("net.Dial(unix,  %s): %w", socket.path, err)
+			}
+			go func() {
+				<-ctx.Done()
+				_ = c.Close()
```
Comment (3Hren) [inline controlplane/modules/route/internal/discovery/bird/export.go:77]: I'd rather log error here as a warning.

File: controlplane/modules/route/internal/discovery/bird/export.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+			parser := NewParser(reader, socket.bufSize, m.log)
+			for {
+				update, err := parser.Next()
+				if err != nil {
+					if err == ErrUnsupportedRDType {
+						// FIXME add telemetry
+						continue
+					}
+					cancel(err)
+					return fmt.Errorf("bird export parser.Next: %w", err)
```
Comment (3Hren) [inline controlplane/modules/route/internal/discovery/bird/export.go:89]: "failed to parse bird chunk: %w" or something more human-readable?

File: controlplane/modules/route/internal/discovery/bird/export.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+				}
+				route := rib.MakeBirdRoute()
+				if err := update.Decode(route); err != nil {
+					cancel(err)
+					return fmt.Errorf("update.Decode(): %w", err)
```
Comment (3Hren) [inline controlplane/modules/route/internal/discovery/bird/export.go:94]: "failed to update route: %w" or something more human-readable?

File: controlplane/modules/route/internal/discovery/bird/testdata/fuzz/Fuzz_update_Decode/0aa332e51f5629d7 [hunk 8]
```diff
@@ -0,0 +1,2 @@
+go test fuzz v1
```
Comment (3Hren) [inline controlplane/modules/route/internal/discovery/bird/testdata/fuzz/Fuzz_update_Decode/0aa332e51f5629d7:1]: I think we don't want to commit fuzzer artifacts. Why do we want to?

File: controlplane/modules/route/internal/discovery/neigh/entry.go [hunk 9]
```diff
@@ -11,13 +12,22 @@ import (
 type NeighbourEntry struct {
 	// NextHop is the IP address of the next hop.
 	NextHop netip.Addr
-	// LinkAddr is the MAC address of the next hop.
-	LinkAddr net.HardwareAddr
-	// HardwareAddr is the MAC address of the local interface that observed
-	// the neighbour.
-	HardwareAddr net.HardwareAddr
+	HardwareRoute
```
Comment (3Hren) [inline controlplane/modules/route/internal/discovery/neigh/entry.go:15]: Let's leave explicit fields - it's more readable. But you can change their type to [6] byte.

I personally don't honor embedded structs, because there are too much magic. For example this struct now have `String()` method.

File: controlplane/modules/route/internal/discovery/neigh/neigh.go [hunk 10]
```diff
@@ -207,17 +207,19 @@ func (m *NeighMonitor) updateNeighbours() error {
...
+			UpdatedAt: time.Now(),
+			State:     NeighbourState(neigh.State),
 		}
 
 		m.log.Debugw("resolved neighbour entry",
 			zap.Stringer("nexthop_addr", entry.NextHop),
-			zap.Stringer("nexthop_hardware_addr", entry.LinkAddr),
-			zap.Stringer("hardware_addr", entry.HardwareAddr),
+			zap.Stringer("nexthop_hardware_addr", neigh.HardwareAddr),
+			zap.Stringer("inface_hardware_addr", hardwareAddr),
```
Comment (3Hren) [inline controlplane/modules/route/internal/discovery/neigh/neigh.go:222]: What was wrong with previous attribute?

File: controlplane/pkg/yncp/cfg.go [hunk 11]
```diff
@@ -20,14 +20,26 @@ type Config struct {
 	Modules ModulesConfig `json:"modules" yaml:"modules"`
 }
 
+func DefaultConfig() *Config {
+	return &Config{
+		Logging: LoggingConfig{
+			Level: zapcore.DebugLevel,
```
Comment (3Hren) [inline controlplane/pkg/yncp/cfg.go:26]: Suggest Info as a reasonable default.

File: controlplane/modules/route/internal/rib/routes.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	Data1 uint32
+	Data2 uint32
+}
+
+type LargeCommunityList struct {
+	LargeCommunity
+	Next *LargeCommunityList
+}
+
+type Route struct {
```
Comment (3Hren) [inline controlplane/modules/route/internal/rib/routes.go:28]: This is an important structure. Please add a description for each field, what it means and where it is important.

File: controlplane/modules/route/mod.go [hunk 13]
```diff
@@ -119,7 +127,7 @@ func (m *RouteModule) Run(ctx context.Context) error {
 	}
 
 	for {
-		if _, err := gateway.Register(ctx, req); err == nil {
+		if _, err = gateway.Register(ctx, req); err == nil {
```
Comment (3Hren) [inline controlplane/modules/route/mod.go:130]: Why?

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: controlplane/modules/route/cfg.go (hunk 1 above)
Comment (sakateka) [reply controlplane/modules/route/cfg.go:20]: By default, nothing happens; instead, an informative log message will be printed to the console.
```
m.log.Info("bird export reader is disabled")
```

File: controlplane/modules/route/internal/discovery/bird/export.go (hunk 4 above)
Comment (sakateka) [reply controlplane/modules/route/internal/discovery/bird/export.go:57]: I think any value greater than zero will be enough here. I will change it to 10, but I don't like the idea of moving it to the config, as it would overwhelm the user without providing significant improvements to system stability.

File: controlplane/modules/route/internal/discovery/neigh/neigh.go (hunk 10 above)
Comment (sakateka) [reply controlplane/modules/route/internal/discovery/neigh/neigh.go:222]: There is too much hardware around; we need some way to distinguish between them. :)

File: controlplane/modules/route/mod.go (hunk 13 above)
Comment (sakateka) [reply controlplane/modules/route/mod.go:130]: `gopls` suggests these changes; it shows a warning about variable shadowing.

**Review by 3Hren — APPROVED** [review]

**Review by 3Hren — DISMISSED** [review]

File: controlplane/modules/route/internal/discovery/bird/testdata/fuzz/Fuzz_update_Decode/0aa332e51f5629d7 (hunk 8 above)
Comment (sakateka) [reply controlplane/modules/route/internal/discovery/bird/testdata/fuzz/Fuzz_update_Decode/0aa332e51f5629d7:1]: Running `go test ./...` will execute the fuzz tests with these artifacts as regular test cases, which can be helpful if the bug reappears. Therefore, it seems simpler and more reliable to commit the fuzz artifacts rather than rewrite this test by hand.

**Review by 3Hren — APPROVED** [review]
